File: java/common/src/test/java/org/apache/shindig/common/uri/UriTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.shindig.common.uri;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;

import org.junit.Test;
import org.junit.Ignore;

import java.net.URI;
import java.util.Arrays;
import java.util.Collection;

/**
 * Tests for Uri.
 */
public class UriTest {
  @Test
  public void parseFull() {
    Uri uri = Uri.parse("http://apache.org/foo?a=b&a=c&b=d+e#blah");

    assertEquals("http", uri.getScheme());
    assertEquals("apache.org", uri.getAuthority());
    assertEquals("/foo", uri.getPath());
    assertEquals("a=b&a=c&b=d+e", uri.getQuery());
    Collection<String> params = Arrays.asList("b", "c");
    assertEquals(params, uri.getQueryParameters("a"));
    assertEquals("b", uri.getQueryParameter("a"));
    assertEquals("d e", uri.getQueryParameter("b"));
    assertEquals("blah", uri.getFragment());

    assertEquals("http://apache.org/foo?a=b&a=c&b=d+e#blah", uri.toString());
  }

  @Test
  public void parseNoScheme() {
    Uri uri = Uri.parse("//apache.org/foo?a=b&a=c&b=d+e#blah");

    assertNull(uri.getScheme());
    assertEquals("apache.org", uri.getAuthority());
    assertEquals("/foo", uri.getPath());
    assertEquals("a=b&a=c&b=d+e", uri.getQuery());
    Collection<String> params = Arrays.asList("b", "c");
    assertEquals(params, uri.getQueryParameters("a"));
    assertEquals("b", uri.getQueryParameter("a"));
    assertEquals("d e", uri.getQueryParameter("b"));
    assertEquals("blah", uri.getFragment());
  }

  @Test
  public void parseNoAuthority() {
    Uri uri = Uri.parse("http:/foo?a=b&a=c&b=d+e#blah");

    assertEquals("http", uri.getScheme());
    assertNull(uri.getAuthority());
    assertEquals("/foo", uri.getPath());
    assertEquals("a=b&a=c&b=d+e", uri.getQuery());
    Collection<String> params = Arrays.asList("b", "c");
    assertEquals(params, uri.getQueryParameters("a"));
    assertEquals("b", uri.getQueryParameter("a"));
    assertEquals("d e", uri.getQueryParameter("b"));
    assertEquals("blah", uri.getFragment());
  }

  @Test
  public void parseNoPath() {
    Uri uri = Uri.parse("http://apache.org?a=b&a=c&b=d+e#blah");

    assertEquals("http", uri.getScheme());
    assertEquals("apache.org", uri.getAuthority());
    // Path is never null.
    assertEquals("", uri.getPath());
    assertEquals("a=b&a=c&b=d+e", uri.getQuery());
    Collection<String> params = Arrays.asList("b", "c");
    assertEquals(params, uri.getQueryParameters("a"));
    assertEquals("b", uri.getQueryParameter("a"));
    assertEquals("d e", uri.getQueryParameter("b"));
    assertEquals("blah", uri.getFragment());
  }

  @Test
  public void parseNoQuery() {
    Uri uri = Uri.parse("http://apache.org/foo#blah");

    assertEquals("http", uri.getScheme());
    assertEquals("apache.org", uri.getAuthority());
    assertEquals("/foo", uri.getPath());
    assertNull(uri.getQuery());
    assertEquals(0, uri.getQueryParameters().size());
    assertNull(uri.getQueryParameter("foo"));
    assertEquals("blah", uri.getFragment());
  }

  @Test
  public void parseNoFragment() {
    Uri uri = Uri.parse("http://apache.org/foo?a=b&a=c&b=d+e");

    assertEquals("http", uri.getScheme());
    assertEquals("apache.org", uri.getAuthority());
    assertEquals("/foo", uri.getPath());
    assertEquals("a=b&a=c&b=d+e", uri.getQuery());
    assertNull(uri.getFragment());
  }

  @Test(expected = IllegalArgumentException.class)
  public void parseInvalidHost() {
    Uri.parse("http://A&E%#%#%/foo?a=b#blah");
  }

  @Test(expected = IllegalArgumentException.class)
  public void parseInvalidScheme() {
    Uri.parse("----://apache.org/foo?a=b#blah");
  }

  @Test(expected = IllegalArgumentException.class)
  public void parseInvalidPath() {
    Uri.parse("http://apache.org/foo\\---(&%?a=b#blah");
  }

  @Test
  public void toJavaUri() {
    URI javaUri = URI.create("http://example.org/foo/bar/baz?blah=blah#boo");
    Uri uri = Uri.parse("http://example.org/foo/bar/baz?blah=blah#boo");

    assertEquals(javaUri, uri.toJavaUri());
  }

  @Test
  public void toJavaUriWithSpecialChars() {
    URI javaUri = URI.create("http://example.org/foo/bar/baz?blah=bl%25ah#boo");
    Uri uri = Uri.parse("http://example.org/foo/bar/baz?blah=bl%25ah#boo");

    assertEquals(javaUri, uri.toJavaUri());
  }

  @Test
  public void fromJavaUri() throws Exception {
    URI javaUri = URI.create("http://example.org/foo/bar/baz?blah=blah#boo");
    Uri uri = Uri.parse("http://example.org/foo/bar/baz?blah=blah#boo");

    assertEquals(uri, Uri.fromJavaUri(javaUri));
  }

  @Test
  public void resolveFragment() throws Exception {
    Uri base = Uri.parse("http://example.org/foo/bar/baz?blah=blah#boo");
    Uri other = Uri.parse("#bar");

    assertEquals("http://example.org/foo/bar/baz?blah=blah#bar", base.resolve(other).toString());
  }

  @Test
  public void resolveQuery() throws Exception {
    Uri base = Uri.parse("http://example.org/foo/bar/baz?blah=blah#boo");
    Uri other = Uri.parse("?hello=world");

    assertEquals("http://example.org/foo/bar/?hello=world", base.resolve(other).toString());
  }

  @Test
  public void resolvePathIncludesSubdirs() throws Exception {
    Uri base = Uri.parse("http://example.org/foo/bar/baz?blah=blah#boo");
    Uri other = Uri.parse("fez/../huey/./dewey/../louis");

    assertEquals("http://example.org/foo/bar/huey/louis", base.resolve(other).toString());
  }

  // Ignore for now..
  @Ignore
  public void resolvePathSubdirsExtendsBeyondRoot() throws Exception {
    Uri base = Uri.parse("http://example.org/foo/bar/baz?blah=blah#boo");
    Uri other = Uri.parse("../random/../../../../../home");

    assertEquals("http://example.org/home", base.resolve(other).toString());
  }

  @Test
  public void resolvePathRelative() throws Exception {
    Uri base = Uri.parse("http://example.org/foo/bar/baz?blah=blah#boo");
    Uri other = Uri.parse("wee");

    assertEquals("http://example.org/foo/bar/wee", base.resolve(other).toString());
  }

  @Test
  public void resolvePathRelativeToNullPath() throws Exception {
    Uri base = new UriBuilder().setScheme("http").setAuthority("example.org").toUri();
    Uri other = Uri.parse("dir");

    assertEquals("http://example.org/dir", base.resolve(other).toString());
  }

  @Test
  public void resolvePathAbsolute() throws Exception {
    Uri base = Uri.parse("http://example.org/foo/bar/baz?blah=blah#boo");
    Uri other = Uri.parse("/blah");

    assertEquals("http://example.org/blah", base.resolve(other).toString());
  }

  @Test
  public void resolveAuthority() throws Exception {
    Uri base = Uri.parse("https://example.org/foo/bar/baz?blah=blah#boo");
    Uri other = Uri.parse("//example.com/blah");

    assertEquals("https://example.com/blah", base.resolve(other).toString());
  }

  @Test
  public void resolveAbsolute() throws Exception {
    Uri base = Uri.parse("http://example.org/foo/bar/baz?blah=blah#boo");
    Uri other = Uri.parse("http://www.ietf.org/rfc/rfc2396.txt");

    assertEquals("http://www.ietf.org/rfc/rfc2396.txt", base.resolve(other).toString());
  }

  @Test
  public void absoluteUrlIsAbsolute() {
    assertTrue("Url with scheme not reported absolute.",
        Uri.parse("http://example.org/foo").isAbsolute());
  }

  @Test
  public void relativeUrlIsNotAbsolute() {
    assertFalse("Url without scheme reported absolute.",
        Uri.parse("//example.org/foo").isAbsolute());
  }

  @Test
  public void parseWithSpecialCharacters() {
    String original = "http://example.org/?foo%25pbar=baz+blah";

    assertEquals(original, Uri.parse(original).toString());
  }

  @Test
  public void equalsAndHashCodeOk() {
    Uri uri = Uri.parse("http://example.org/foo/bar/baz?blah=blah#boo");
    Uri uri2 = new UriBuilder()
        .setScheme("http")
        .setAuthority("example.org")
        .setPath("/foo/bar/baz")
        .addQueryParameter("blah", "blah")
        .setFragment("boo")
        .toUri();

    assertEquals(uri, uri2);
    assertEquals(uri2, uri);

    assertEquals(uri.hashCode(), uri2.hashCode());
  }
}

File: 2.0.x/features/src/test/javascript/features/shindig.uri/uritest.js
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/**
 * @fileoverview
 *
 * Unittests for the shindig.uri library.
 */

function ShindigUriTest(name) {
  TestCase.call(this, name);
}

ShindigUriTest.inherits(TestCase);

ShindigUriTest.prototype.testParseFullUri = function() {
  var str = "http://www.example.com/my/path?qk1=qv1&qk2=qv2#fk1=fv1&fk2=fv2";
  var uri = shindig.uri(str);

  this.assertEquals("http", uri.getSchema());
  this.assertEquals("www.example.com", uri.getAuthority());
  this.assertEquals("http://www.example.com", uri.getOrigin());
  this.assertEquals("/my/path", uri.getPath());
  this.assertEquals("qk1=qv1&qk2=qv2", uri.getQuery());
  this.assertEquals("fk1=fv1&fk2=fv2", uri.getFragment());
  this.assertEquals("qv1", uri.getQP("qk1"));
  this.assertEquals("qv2", uri.getQP("qk2"));
  this.assertEquals("fv1", uri.getFP("fk1"));
  this.assertEquals("fv2", uri.getFP("fk2"));

  // Check query and fragment again to ensure ordering doesn't matter.
  this.assertEquals("qk1=qv1&qk2=qv2", uri.getQuery());
  this.assertEquals("fk1=fv1&fk2=fv2", uri.getFragment());

  // Re-serialize all.
  this.assertEquals(str, uri.toString());
};

ShindigUriTest.prototype.testParseQuerylessUri = function() {
  var str = "http://www.example.com/my/path#fk1=fv1&fk2=fv2";
  var uri = shindig.uri(str);

  this.assertEquals("http", uri.getSchema());
  this.assertEquals("www.example.com", uri.getAuthority());
  this.assertEquals("http://www.example.com", uri.getOrigin());
  this.assertEquals("/my/path", uri.getPath());
  this.assertEquals("fk1=fv1&fk2=fv2", uri.getFragment());
  this.assertEquals("fv1", uri.getFP("fk1"));
  this.assertEquals("fv2", uri.getFP("fk2"));

  // Check query and fragment again to ensure ordering doesn't matter.
  this.assertEquals("fk1=fv1&fk2=fv2", uri.getFragment());

  // Re-serialize all.
  this.assertEquals(str, uri.toString());
};

ShindigUriTest.prototype.testParseFragmentlessUri = function() {
  var str = "http://www.example.com/my/path?qk1=qv1&qk2=qv2";
  var uri = shindig.uri(str);

  this.assertEquals("http", uri.getSchema());
  this.assertEquals("www.example.com", uri.getAuthority());
  this.assertEquals("http://www.example.com", uri.getOrigin());
  this.assertEquals("/my/path", uri.getPath());
  this.assertEquals("qk1=qv1&qk2=qv2", uri.getQuery());
  this.assertEquals("qv1", uri.getQP("qk1"));
  this.assertEquals("qv2", uri.getQP("qk2"));

  // Check query and fragment again to ensure ordering doesn't matter.
  this.assertEquals("qk1=qv1&qk2=qv2", uri.getQuery());

  // Re-serialize all.
  this.assertEquals(str, uri.toString());
};

ShindigUriTest.prototype.testParseSchemalessUri = function() {
  var str = "//www.example.com/my/path?qk1=qv1&qk2=qv2#fk1=fv1&fk2=fv2";
  var uri = shindig.uri(str);

  this.assertEquals("", uri.getSchema());
  this.assertEquals("www.example.com", uri.getAuthority());
  this.assertEquals("//www.example.com", uri.getOrigin());
  this.assertEquals("/my/path", uri.getPath());
  this.assertEquals("qk1=qv1&qk2=qv2", uri.getQuery());
  this.assertEquals("fk1=fv1&fk2=fv2", uri.getFragment());
  this.assertEquals("qv1", uri.getQP("qk1"));
  this.assertEquals("qv2", uri.getQP("qk2"));
  this.assertEquals("fv1", uri.getFP("fk1"));
  this.assertEquals("fv2", uri.getFP("fk2"));

  // Check query and fragment again to ensure ordering doesn't matter.
  this.assertEquals("qk1=qv1&qk2=qv2", uri.getQuery());
  this.assertEquals("fk1=fv1&fk2=fv2", uri.getFragment());

  // Re-serialize all.
  this.assertEquals(str, uri.toString());
};

ShindigUriTest.prototype.testParseAuthoritylessUri = function() {
  var str = "/my/path?qk1=qv1&qk2=qv2#fk1=fv1&fk2=fv2";
  var uri = shindig.uri(str);

  this.assertEquals("", uri.getSchema());
  this.assertEquals("", uri.getAuthority());
  this.assertEquals("", uri.getOrigin());
  this.assertEquals("/my/path", uri.getPath());
  this.assertEquals("qk1=qv1&qk2=qv2", uri.getQuery());
  this.assertEquals("fk1=fv1&fk2=fv2", uri.getFragment());
  this.assertEquals("qv1", uri.getQP("qk1"));
  this.assertEquals("qv2", uri.getQP("qk2"));
  this.assertEquals("fv1", uri.getFP("fk1"));
  this.assertEquals("fv2", uri.getFP("fk2"));

  // Check query and fragment again to ensure ordering doesn't matter.
  this.assertEquals("qk1=qv1&qk2=qv2", uri.getQuery());
  this.assertEquals("fk1=fv1&fk2=fv2", uri.getFragment());

  // Re-serialize all.
  this.assertEquals(str, uri.toString());
};

ShindigUriTest.prototype.testParsePathlessUri = function() {
  var str = "http://www.example.com?qk1=qv1&qk2=qv2#fk1=fv1&fk2=fv2";
  var uri = shindig.uri(str);

  this.assertEquals("http", uri.getSchema());
  this.assertEquals("www.example.com", uri.getAuthority());
  this.assertEquals("http://www.example.com", uri.getOrigin());
  this.assertEquals("qk1=qv1&qk2=qv2", uri.getQuery());
  this.assertEquals("fk1=fv1&fk2=fv2", uri.getFragment());
  this.assertEquals("qv1", uri.getQP("qk1"));
  this.assertEquals("qv2", uri.getQP("qk2"));
  this.assertEquals("fv1", uri.getFP("fk1"));
  this.assertEquals("fv2", uri.getFP("fk2"));

  // Check query and fragment again to ensure ordering doesn't matter.
  this.assertEquals("qk1=qv1&qk2=qv2", uri.getQuery());
  this.assertEquals("fk1=fv1&fk2=fv2", uri.getFragment());

  // Re-serialize all.
  this.assertEquals(str, uri.toString());
};

ShindigUriTest.prototype.testParsePathOnly = function() {
  var str = "/my/path";
  var uri = shindig.uri(str);

  this.assertEquals("", uri.getSchema());
  this.assertEquals("", uri.getAuthority());
  this.assertEquals("", uri.getOrigin());
  this.assertEquals("/my/path", uri.getPath());

  this.assertEquals(str, uri.toString());
};

ShindigUriTest.prototype.testParseQueryOnly = function() {
  var str = "?foo=bar&baz=bak&boo=hiss";
  var uri = shindig.uri(str);

  this.assertEquals("bar", uri.getQP("foo"));
  this.assertEquals("bak", uri.getQP("baz"));
  this.assertEquals("hiss", uri.getQP("boo"));

  this.assertEquals(str, uri.toString());
};

ShindigUriTest.prototype.testParseFragmentOnly = function() {
  var str = "#foo=bar&baz=bak";
  var uri = shindig.uri(str);

  this.assertEquals("bar", uri.getFP("foo"));
  this.assertEquals("bak", uri.getFP("baz"));

  this.assertEquals(str, uri.toString());
};

ShindigUriTest.prototype.testParseWithContextualOddities = function() {
  var uri = shindig.uri("//www.example.com/my//path?#");

  this.assertEquals("", uri.getSchema());
  this.assertEquals("www.example.com", uri.getAuthority());
  this.assertEquals("/my//path", uri.getPath());
  this.assertEquals("", uri.getQuery());
  this.assertEquals("", uri.getFragment());

  this.assertEquals("//www.example.com/my//path", uri.toString());
};

ShindigUriTest.prototype.testParseQueryNullAndMissing = function() {
  var uri = shindig.uri("?&one=two&three&&four=five");
  this.assertEquals("two", uri.getQP("one"));
  this.assertTrue(null === uri.getQP("three"));
  this.assertTrue(undefined === uri.getQP("nonexistent"));
  this.assertEquals("five", uri.getQP("four"));
};

ShindigUriTest.prototype.testParseFragmentNullAndMissing = function() {
  var uri = shindig.uri("#&one=two&three&&four=five");
  this.assertEquals("two", uri.getFP("one"));
  this.assertTrue(null === uri.getFP("three"));
  this.assertTrue(undefined === uri.getQP("nonexistent"));
  this.assertEquals("five", uri.getFP("four"));
};

ShindigUriTest.prototype.testBuildFullUri = function() {
  var uri = shindig.uri().setSchema("http")
                         .setAuthority("www.example.com")
                         .setPath("/my/path")
                         .setQuery("?one=two&three=four")
                         .setFragment("#five=six");
  this.assertEquals("http://www.example.com/my/path?one=two&three=four#five=six", uri.toString());
};

ShindigUriTest.prototype.testBuildSchemalessUri = function() {
  var uri = shindig.uri().setAuthority("www.example.com")
                         .setPath("/my/path")
                         .setQuery("?one=two&three=four")
                         .setFragment("#five=six");
  this.assertEquals("//www.example.com/my/path?one=two&three=four#five=six", uri.toString());
};

ShindigUriTest.prototype.testBuildAuthoritylessUri = function() {
  var uri = shindig.uri().setPath("/my/path")
                         .setQuery("?one=two&three=four")
                         .setFragment("#five=six");
  this.assertEquals("/my/path?one=two&three=four#five=six", uri.toString());
};

ShindigUriTest.prototype.testBuildPathlessUri = function() {
  var uri = shindig.uri().setSchema("http")
                         .setAuthority("www.example.com")
                         .setQuery("?one=two&three=four")
                         .setFragment("#five=six");
  this.assertEquals("http://www.example.com?one=two&three=four#five=six", uri.toString());
};

ShindigUriTest.prototype.testBuildQuerylessUri = function() {
  var uri = shindig.uri().setSchema("http")
                         .setAuthority("www.example.com")
                         .setPath("/my/path")
                         .setFragment("#five=six");
  this.assertEquals("http://www.example.com/my/path#five=six", uri.toString());
};

ShindigUriTest.prototype.testBuildFragmentlessUri = function() {
  var uri = shindig.uri().setSchema("http")
                         .setAuthority("www.example.com")
                         .setPath("/my/path")
                         .setQuery("?one=two&three=four");
  this.assertEquals("http://www.example.com/my/path?one=two&three=four", uri.toString());
};

ShindigUriTest.prototype.testBuildPath = function() {
  var uri = shindig.uri().setPath("/my/path");
  this.assertEquals("/my/path", uri.toString());
};

ShindigUriTest.prototype.testBuildAuthority = function() {
  var uri = shindig.uri().setAuthority("www.example.com");
  this.assertEquals("//www.example.com", uri.toString());
};

ShindigUriTest.prototype.testBuildDirectQuery = function() {
  var uri = shindig.uri().setQuery("one=two&three&&four=five");
  this.assertEquals("two", uri.getQP("one"));
  this.assertTrue(null === uri.getQP("three"));
  this.assertTrue(undefined === uri.getQP("nonexistent"));
  this.assertEquals("five", uri.getQP("four"));
};

ShindigUriTest.prototype.testBuildDirectFragment = function() {
  var uri = shindig.uri().setFragment("one=two&three&&four=five");
  this.assertEquals("two", uri.getFP("one"));
  this.assertTrue(null === uri.getFP("three"));
  this.assertTrue(undefined === uri.getQP("nonexistent"));
  this.assertEquals("five", uri.getFP("four"));
};

ShindigUriTest.prototype.testBuildQuery = function() {
  var uri = shindig.uri().setQuery("one=two&three=four");
  this.assertEquals("two", uri.getQP("one"));
  this.assertEquals("four", uri.getQP("three"));
  uri.setQP("three", "five");
  this.assertEquals("two", uri.getQP("one"));
  this.assertEquals("five", uri.getQP("three"));
  uri.setQP({ one: "one", two: "two", three: "three" });
  this.assertEquals("one", uri.getQP("one"));
  this.assertEquals("two", uri.getQP("two"));
  this.assertEquals("three", uri.getQP("three"));
  uri.setQP("two", null);
  this.assertEquals("?one=one&three=three&two", uri.toString());
};

ShindigUriTest.prototype.testBuildFragment = function() {
  var uri = shindig.uri().setFragment("one=two&three=four");
  this.assertEquals("two", uri.getFP("one"));
  this.assertEquals("four", uri.getFP("three"));
  uri.setFP("three", "five");
  this.assertEquals("two", uri.getFP("one"));
  this.assertEquals("five", uri.getFP("three"));
  uri.setFP({ one: "one", two: "two", three: "three" });
  this.assertEquals("one", uri.getFP("one"));
  this.assertEquals("two", uri.getFP("two"));
  this.assertEquals("three", uri.getFP("three"));
  uri.setFP("two", null);
  this.assertEquals("#one=one&three=three&two", uri.toString());
};

ShindigUriTest.prototype.testReplaceExistingQuery = function() {
  var uri = shindig.uri().setQuery("one=two")
                         .setFragment("three=four")
                         .setExistingP("one", "111")
                         .setExistingP("three", "333")
                         .setExistingP("xxx", "yyy");
  this.assertEquals("?one=111#three=333", uri.toString());
};

ShindigUriTest.prototype.testBuildWithOverrides = function() {
  var uri =
      shindig.uri("http://www.example.com/my/path?one=two&baz#three=four")
        .setAuthority("www.foo.com")
        .setQP("one", "five")
        .setFragment("foo=bar");
  this.assertEquals(null, uri.getQP("baz"));
  this.assertEquals(uri.toString(), "http://www.foo.com/my/path?one=five&baz#foo=bar", uri.toString());
};
